### PR TITLE
Add `DeletedWith` resource option

### PR DESCRIPTION
**Pull request by same-id**
# Description

In many cases there is no need to delete resources if the container resource is going to be deleted as well.

A few examples:
 * Database object (roles, tables) when database is being deleted
 * Cloud IAM bindings when user itself is being deleted

This helps with:
 * Speeding the deletion process
 * Removing unnecessary calls to providers
 * Avoiding failed deletions when the pulumi user running the plan has access to the container resource but not the contained ones

To avoid deleting contained resources, set the `DeletedWith` resource option to the container resource.

TODO:
 Should we support DeletedWith with PendingDeletes?
 Special case might be when contained resource is marked as pending deletion
 but we now want to delete the container resource, so ultimately there is no
 need to delete the contained anymore

Fixes # (issue)

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

**Comment by Frassle** [comment]
> In many cases there is no need to delete resources if the container resource is going to be deleted as well.

We have an issue for this https://github.com/pulumi/pulumi/issues/10634.

I'm not sure "parent" is the right relationship to be tracking this, it already is suffering a bit from confusing and overloaded semantics. 

My original thinking about this was that users wouldn't be able to control it at all, instead resource providers would just state the container relationships. On further thinking that probably won't suffice, if just because providers won't always be fully up to date and people will need local workarounds.

But any option we do add should synchronize well with the idea that providers might also be setting things here, and as above I don't think it should have anything to do with the parent relationship.

**Comment by sam-utila** [comment]
> > In many cases there is no need to delete resources if the container resource is going to be deleted as well.
> 
> We have an issue for this #10634.

Dor from our company (utila.io) opened it :-) 

> 
> I'm not sure "parent" is the right relationship to be tracking this, it already is suffering a bit from confusing and overloaded semantics.
> 

I was also not sure about using Parent for this feature since it binds the child resource to inherit properties (like transformations) that might not be relevant for "skip deletion" functionality.

Another option was to use just a `URN DeletedWith` resource option that points to the container resource instead of piggybacking on top of existing Parent field just because it can specify a `URN`.
Although from naming perspective I was at first sure that the Parent option was the thing I was looking for and would in fact give us this functionality (before this change).

> My original thinking about this was that users wouldn't be able to control it at all, instead resource providers would just state the container relationships. On further thinking that probably won't suffice, if just because providers won't always be fully up to date and people will need local workarounds.
> 

I agree, probably we can model a `DeletedWith` relation between multiple resources, accept the fact that some resources have a mutual `DeletedWith` relation, or maybe even circular one.
In fact, a resource might have multiple `DeletedWith` relations: (hypothetical example) An edge (x, y) is deleted automatically if either x or y are deleted.

Workaround might be to extend with `ExtraDeletedWith` as an array of resources. (just in terms of naming)

Maybe even if resources are "skipping" deletion we would also want to communicate that to the user as a different operation: "delegated delete" instead of "delete"

> But any option we do add should synchronize well with the idea that providers might also be setting things here, and as above I don't think it should have anything to do with the parent relationship.

I would love the overriding functionality hit master from one side since this solves an acute pain today.

From the other hand I do understand that some changes deserve some time for discussion and to "sleep on it" - and then of course spend some time to pick the perfect option name :-)

However I do think that I would hate to see this feature being dragged on out of existance.

As I mentioned (in the TODO:) there are some edge cases like "Pending Deletions" + current deletions that we might also dwell on, or cases where the providers are changing: GCP provider -> Postgres Cloud SQL DB -> Postgres provider -> Role; deleting the Postgres Cloud SQL DB should save the deletion of the Role within it, even though they are handled by different providers.
So waiting out for the perfect solution might take some time (as I assume this feature is something that was probably discussed very early on on Pulumi HQ's whiteboards but was probably, and justifiably put off in favor of simplicity)

Maybe if we define this feature under non-stable API/ABI and release it under PULUMI_EXPERIMENTAL=1 (or a specific experimental flag, to avoid getting ALL experimental features) it will allow us to release this feature faster for people that like to live on the edge :-)

Thanks!

**Comment by Frassle** [comment]
> Dor from our company (utila.io) opened it :-)

👍  

> I would love the overriding functionality hit master from one side since this solves an acute pain today.

Sure, I'm quite happy to just get a new resource option in now and we can internally iterate on later extending schemas and things to explain these relationships as well. The engine is not in place to be able to use that information yet, so it would just be wasted effort if done right now.

> Another option was to use just a URN DeletedWith resource option that points to the container resource instead of piggybacking on top of existing Parent field just because it can specify a URN.

I think something like this is much more workable.

> In fact, a resource might have multiple DeletedWith relations: (hypothetical example) An edge (x, y) is deleted automatically if either x or y are deleted.

Is that likely? I've been trying to think of any examples where a resource has multiple containers, where if either any or all get deleted it also gets deleted but so far everything I've thought of only has one container. My worry with adding a `URN list` option is you have to decided if that's an AND or OR operation (i.e. if `x` has `deletedWith: [y, z]` do you delete `x` if just `y` is deleted or only when `y` and `z` are deleted).

> Maybe even if resources are "skipping" deletion we would also want to communicate that to the user as a different operation: "delegated delete" instead of "delete"

I think that will be needed. The engine still needs to do a state delete, and if we ever add lifecycle delete hooks they would still need to fire, so this isn't skipping the delete from the engines point of view, just from the providers point of view.

**Comment by iwahbe** [comment]
> Sure, I'm quite happy to just get a new resource option in now and we can internally iterate on later extending schemas and things to explain these relationships as well. The engine is not in place to be able to use that information yet, so it would just be wasted effort if done right now.

We have precedent for this strategy. `replaceOnChanges` is implemented as a resource option, which we later added to the schema. The schema communicates to the engine via embedding the provided `replaceOnChanges` values in the generated SDK, which passes them to the engine.

**Comment by same-id** [comment]
> > Dor from our company (utila.io) opened it :-)
> 
> 👍
> 
> > I would love the overriding functionality hit master from one side since this solves an acute pain today.
> 
> Sure, I'm quite happy to just get a new resource option in now and we can internally iterate on later extending schemas and things to explain these relationships as well. The engine is not in place to be able to use that information yet, so it would just be wasted effort if done right now.
> 
> > Another option was to use just a URN DeletedWith resource option that points to the container resource instead of piggybacking on top of existing Parent field just because it can specify a URN.
> 
> I think something like this is much more workable.
> 

Tell me if you want me to resubmit the PR with that option (and maybe choose a different name) or whether you are going to implement this internally in Pulumi

> > In fact, a resource might have multiple DeletedWith relations: (hypothetical example) An edge (x, y) is deleted automatically if either x or y are deleted.
> 
> Is that likely? I've been trying to think of any examples where a resource has multiple containers, where if either any or all get deleted it also gets deleted but so far everything I've thought of only has one container. My worry with adding a `URN list` option is you have to decided if that's an AND or OR operation (i.e. if `x` has `deletedWith: [y, z]` do you delete `x` if just `y` is deleted or only when `y` and `z` are deleted).
> 

I thought of OR relation only.
My initial thought was something like in GCP you can create a new role and a new user, and then bind that role in IAM.

If the role is deleted (need to make sure) then the IAM binding is deleted
If the user is deleted then the IAM binding is deleted

So:

IAM-Binding { deletedWith: [NewRole, NewUser] } (no need to delete the IAM binding if role OR user is being deleted)

But need to make sure.

> > Maybe even if resources are "skipping" deletion we would also want to communicate that to the user as a different operation: "delegated delete" instead of "delete"
> 
> I think that will be needed. The engine still needs to do a state delete, and if we ever add lifecycle delete hooks they would still need to fire, so this isn't skipping the delete from the engines point of view, just from the providers point of view.

**Comment by Frassle** [comment]
> Tell me if you want me to resubmit the PR with that option (and maybe choose a different name) or whether you are going to implement this internally in Pulumi

If you can I'm happy to help with reviewing that and getting it in.

> IAM-Binding { deletedWith: [NewRole, NewUser] } (no need to delete the IAM binding if role OR user is being deleted)

That sounds reasonable. So if either parent gets deleted the IAM binding would get deleted by aws as well so rather than us doing (Delete binding, Delete role) we can safely just do (Delete role) to get the same end result?

Given that I think a `URN[] DeletedBy` resource option sounds correct?

**Comment by errordeveloper** [comment]
As I mentioned in https://github.com/pulumi/pulumi-kubernetes/issues/2211, Kubernetes is another example. There are, of course, some object that own cloud resources, like service with an external LB, storage volumes and various CRs (perhaps stacks managed by [the Pulumi operator](https://github.com/pulumi/pulumi-kubernetes-operator) would be a good example along with CrossPlane and other similar solutions). The API will need to have an explicit opt-in with a default allow filter (so that user doesn't need to enumerate all the common resource types).

**Comment by same-id** [comment]
> That sounds reasonable. So if either parent gets deleted the IAM binding would get deleted by aws as well so rather than us doing (Delete binding, Delete role) we can safely just do (Delete role) to get the same end result?
> 
> Given that I think a `URN[] DeletedBy` resource option sounds correct?

Not sure about the name `DeletedBy`, "A is deleted if B is deleted" made me choose `DeletedWith`

One more thing that concerns me is that I assume that you can't specify circular `Parent` dependencies.
(Pulumi shows all children as `⌞`)

In the current implementation I used the `DeletedWithParent` which helped. Otherwise, in the current implementation if we specify:
A `DeletedWith` B
B `DeletedWith` A

Then none will get deleted.

WDYT

**Comment by Frassle** [comment]
> One more thing that concerns me is that I assume that you can't specify circular Parent dependencies.

That's correct for Parents, and also for Dependencies.

> A `DeletedWith` B
> B `DeletedWith` A
>
> Then none will get deleted.

It's not generally possible to define circular resources in pulumi programs anyway, but given that `x DeletedWith y` tells the engine that if it deletes `y` then it can elide the delete for `x`, it doesn't define the deletion order (that's done via dependencies). So if you change the program to just delete A or B in the above case it will just elide the actual delete for the other but mark it as deleted. If you delete both the engine will decide which one to really delete based on dependency ordering (which can't be circular). So even if we did end up with cycles in the DeletedWith graph I think the engine could cope with it.

**Comment by same-id** [comment]
I understand, I was referring to my current implementation:

Now, I skip deletion if `DeletedWithParent` is in `otherDeletions`==`sg.deletes`

Where `sg.deletes` is all the deletions that are about to happen.

If I change `DeletedWithParent` to `DeletedWith` and then supply:
A deleted with B
B deleted with A

Then both deletions will be skipped, which is not intended.

So there are 3 options:
1. Keep using `DeletedWithParent` and have additional `Parent` semantics
2. Change to `DeletedWith` and not allow cycles (need to make sure) - is this by default? I think we need to add checks
3. Change to `DeletedWith` and allow cycles, but make sure at least one element in the cycle is deleted, it's a bit more tricky and the current `sg.deletes` implementation will not work IMO

**Comment by Frassle** [comment]
> So there are 3 options:

Option three is clearly the ideal, but if it's overly tricky we could take option 2 for the first version of this feature.

**Comment by Frassle** [comment]
> If there's a cycle, none will get deleted, and it's up to the user to avoid this, it's an "expert override" feature anyway, just need to document this.

Yeh actually that's fine for the first pass at this.

**Comment by same-id** [comment]
Refactored `Parent` out.
PTAL

**Review by Frassle — COMMENTED** [review]
We should add this as an engine feature (see https://github.com/same-id/pulumi/blob/c28451d1f5169aae4b10090b483a0a41b9a0249b/pkg/resource/deploy/source_eval.go#L628) and check for that support in SDKs if the deletedWith resource option is set.
This is so that if people update to the latest SDK but don't update their CLI they'll get an error that DeletedWith won't work until their CLI is updated, we forgot to do this for `retainOnDelete` and it caused a lot of confusion (https://github.com/pulumi/pulumi/issues/9255)

File: pkg/resource/deploy/step_generator.go [hunk 1]
```diff
@@ -1242,7 +1243,10 @@ func (sg *stepGenerator) GeneratePendingDeletes() []Step {
 				logging.V(7).Infof(
 					"stepGenerator.GeneratePendingDeletes(): resource (%v, %v) is pending deletion", res.URN, res.ID)
 				sg.pendingDeletes[res] = true
-				dels = append(dels, NewDeleteStep(sg.deployment, res))
+				// TODO: Should we support DeletedWith with PendingDeletes
```
[the remark this answers is not in the document]
Comment (same-id) [reply pkg/resource/deploy/step_generator.go:1246]: So keep the comment and delete it when that logic is removed? Or remove it now?

File: sdk/go/common/apitype/core.go [hunk 2]
```diff
@@ -326,6 +326,9 @@ type ResourceV3 struct {
 	ImportID resource.ID `json:"importID,omitempty" yaml:"importID,omitempty"`
 	// If set to True, the providers Delete method will not be called for this resource. Pulumi simply stops tracking the deleted resource.
 	RetainOnDelete bool `json:"retainOnDelete,omitempty" yaml:"retainOnDelete,omitempty"`
+	// If set, the providers Delete method will not be called for this resource
+	// if specified resource is being deleted as well.
+	DeletedWith resource.URN `json:"deletedWith,omitempty" yaml:"deletedWith,omitempty"`
```
[the remark this answers is not in the document]
Comment (same-id) [reply sdk/go/common/apitype/core.go:331]: Yea, it makes sense

**Comment by same-id** [comment]
> We should add this as an engine feature (see https://github.com/same-id/pulumi/blob/c28451d1f5169aae4b10090b483a0a41b9a0249b/pkg/resource/deploy/source_eval.go#L628) and check for that support in SDKs if the deletedWith resource option is set. This is so that if people update to the latest SDK but don't update their CLI they'll get an error that DeletedWith won't work until their CLI is updated, we forgot to do this for `retainOnDelete` and it caused a lot of confusion (#9255)

It makes sense, however if I understand correctly I need to:
1. Add "true" for feature "deletedWith" at server
2. Check that feature in all SDKs (Go, Python, C#, typescript) but fail only if DeletedWith is not empty for some resource.

I tried to find other existing features that should work like that for a dumb "copy-paste-rename" but couldn't find any, so I would appreciate some help here. 🙇

**Comment by Frassle** [comment]
> I tried to find other existing features that should work like that for a dumb "copy-paste-rename" but couldn't find any, so I would appreciate some help here. 🙇

Sure :) If you look at how we did this for aliases you can copy that:
Here's an example of adding it to source_eval:
https://github.com/pulumi/pulumi/pull/9731/files#diff-984b3279d61b588030474831898bb76d093d863ca2d0015bb081d842602a9ab7R627

And then an example of using that in the dotnet SDK:
Adding it as a feature to lookup: https://github.com/pulumi/pulumi/pull/9731/files#diff-d30dc14f1dbd36376887f48d62eab55cd120eeb1cd157a70fc3c894e553be57cR218
Checking if that feature is set: https://github.com/pulumi/pulumi/pull/9731/files#diff-dd8095fbcc1deacedb06bbbd6c71a7188abe6b1df89811168608a57ee2cef9edR131

Aliases had a back-compat way to still work when the monitor didn't support the new alias spec feature, but in this case you can just throw an exception if MonitorSupportsDeletedWith returns false but DeletedWith is set.

**Comment by same-id** [comment]
I added a new seperate commit for that.
Super not sure about it (not sure how to even test this across different sdks)

File: sdk/dotnet/Pulumi/Deployment/Deployment_RegisterResource.cs [hunk 3]
```diff
@@ -21,6 +21,10 @@ public partial class Deployment
             var label = $"resource:{name}[{type}]";
             Log.Debug($"Registering resource start: t={type}, name={name}, custom={custom}, remote={remote}");
 
+            if (options.DeletedWith != null && await MonitorSupportsDeletedWith().ConfigureAwait(false)) {
```
Comment (Frassle) [inline sdk/dotnet/Pulumi/Deployment/Deployment_RegisterResource.cs:24]: Missing a bool negation here
```suggestion
            if (options.DeletedWith != null && !(await MonitorSupportsDeletedWith().ConfigureAwait(false))) {
```

File: sdk/dotnet/Pulumi/Deployment/Deployment_RegisterResource.cs [hunk 4]
```diff
@@ -21,6 +21,10 @@ public partial class Deployment
...
+                throw new Exception("resource option DeletedWith not supported - please update pulumi")
```
Comment (Frassle) [inline sdk/dotnet/Pulumi/Deployment/Deployment_RegisterResource.cs:25]: Small change to wording to match some other "please update" errors we already have.
```suggestion
                throw new Exception(""The Pulumi CLI does not support the DeletedWith option. Please update the Pulumi CLI."")
```
If you could apply that re-wording to the other SDKs as well please.

**Comment by Frassle** [comment]
> Super not sure about it (not sure how to even test this across different sdks)

Change looks mostly fine, just a missing ! in dotnet, and a small rewording request.

**Comment by same-id** [comment]
Done

**Review by Frassle — APPROVED** [review]
I'm happy with this and no one else has commented anything to the contrary so

bors merge

**Comment by Frassle** [comment]
bors merge

**Comment by Frassle** [comment]
bors retry

**Comment by Frassle** [comment]
bors merge

**Comment by Frassle** [comment]
bors retry

**Comment by Frassle** [comment]
No worries :) Thanks for the substantial contribution!

**Comment by Frassle** [comment]
We do plan on changing that, it will be a breaking change but it's likely no one is using this yet.

**Pull request merged**